waku/node/waku_node.nim

```diff
@@ -104,6 +105,8 @@ type
     announcedAddresses* : seq[MultiAddress]
     started*: bool # Indicates that node has started listening
     topicSubscriptionQueue*: AsyncEventQueue[SubscriptionEvent]
+    handlersTable: Table[string, TopicHandler]
+    subscriptions: Table[PubsubTopic, tuple[pubsubTopic: bool, contentTopics: int]]
```

Comment:
Ok, after thinking about this a bit more, I realise it is because we now have two ways of subscribing - via pubsubTopic directly or via the autosharding-contentTopic route. I would really try to do something simple here, such as assuming that we will never have a mixed configuration: in other words, if someone unsubscribes from a contentTopic we can assume that they didn't _also_ configure/subscribe to the related shard directly, so it should be safe to unsubscribe from the related shard if there are no more handlers. This is a simplifying assumption and introduces a limitation, but I think it's reasonable in order to keep the node as clean as possible. I know there will always be a default handler remaining, but that should be possible to determine from the underlying PubSub object. I don't think we need to track anything on the node.

Reply:
We could just return error messages `not subscribe to this shard` or `shard in use, cannot unsubscribe`

About the limitation, does this mean we can never use `autoSubscribe` if we used `subscribe` before without using `unsubscribe` first?
OR is is per shard, either use pubsub "manual" or content "auto" topic not both?

What about multiple content topic on the same shard? We still need to track the handlers otherwise `autoUnsubscribe` would remove the other content topic handlers.

Reply:
Thinking about it more. To return error messages we still have to track if a "manual" shard subscriptions happened.
It does not reduce the state we have to track at all.

Reply:
No, I would say that the limitation is that, if you have content topic subscriptions, regardless of whether you've also manually subscribed to the same shard, if you unsubscribe from all content topic subscriptions for a shard you'll also be unsubscribed from that shard. In other words, I don't think we need to do anything special: subscribe(contentTopic) adds a handler for the related shard, unsubscribe(contentTopic) removes the associated handler from the shard. If after an unsubscribe(contentTopic) there are no more handlers (except the default), we unsubscribed fully from the shard. Indeed, we'd still need to track the handlers per content topic, but I don't think we'd need to track any other state.

Reply:
This would mean that if you are subscribing/unsubscribing from applications (content topics), you'll have consistent behaviour. If somehow you also want to continue serving the shard after removing all application subscriptions matching that shard, you'd need to do another explicit `subscribe(pubsubTopic)`.

Reply:
Ok make sense!